src/webapp/src/characters/charactersSlice.ts

```diff
@@ -0,0 +1,83 @@
+import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
+
+interface ApiExtraInformation {
+  count: number;
+  pages: number;
+  next: string;
+  prev: string | null;
+}
+
+interface Character {
+  id: number;
+  name: string;
+  status: string;
+  species: string;
+  type: string;
+  gender: string;
+  origin: {
+    name: string;
+    url: string;
+  };
+  location: {
+    name: string;
+    url: string;
+  };
+  episode: string[];
+  url: string;
+  created: Date;
+  image: string;
+}
+
+export interface ApiData {
+  results: Character[];
+  info: ApiExtraInformation;
+  error?: string | null;
+}
+
+const initialState = {
+  results: [] as Character[],
+  info: {} as ApiExtraInformation,
+  error: null,
+} as ApiData;
+
+interface ApiKnownError {
+  error: string;
+}
+
+export const getCharacters = createAsyncThunk<
+  ApiData,
+  void,
+  { rejectValue: ApiKnownError }
+>('characters/getCharacters', async (_, { rejectWithValue }) => {
+  const response = await fetch(`https://rickandmortyapi.com/api/character`);
+
+  if (response.status === 404) {
+    return rejectWithValue((await response.json()) as ApiKnownError);
+  }
+
+  return (await response.json()) as ApiData;
+});
+
+const charactersSlice = createSlice({
+  name: 'characters',
+  initialState,
+  reducers: {},
+  extraReducers: (builder) => {
+    builder.addCase(getCharacters.pending, () => {});
+
+    builder.addCase(getCharacters.rejected, (state, action) => {
+      if (action.payload) {
+        return { ...state, error: action.payload.error };
+      }
+      return { ...state, error: action.error.message };
```

Comment:
Mind explaining to me, when is the case that the code comes down here? I got a little bit confused

Reply:
Of course! Per the documentation:

> if the promise resolved with a rejectWithValue(value) return value, dispatch the rejected action with the value passed into action.payload and 'Rejected' as action.error.message

But:
> if the promise failed and was not handled with `rejectWithValue`, dispatch the rejected action with a serialized version of the error value as action.error

So if the promise failed and we passed the error to `rejectWithValue`, we can get that error from the `action.payload` and we
return `return { ...state, error: action.payload.error }`

But if the promise failed and we didn't rejected it using `rejectWithValue` `action.payload` is going to be `undefined` so we pass the default error that can be found in `action.error.message`

Reply:
The explanation is good. I will have a look onto it and let you know.